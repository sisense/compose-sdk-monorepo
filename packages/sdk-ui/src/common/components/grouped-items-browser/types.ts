export type GroupedItemsBrowserProps = {
  groupedItems: ItemsGroup[];
  groupSecondaryActionConfig?: GroupSecondaryActionConfig;
  itemActionConfig?: ItemActionConfig;
  itemSecondaryActionConfig?: ItemSecondaryActionConfig;
  onScrolledToBottom?: () => void;
};

/**
 * A group of items.
 */
export type ItemsGroup = {
  title: string;
  id: string;
  items: Item[];
  Icon?: React.ComponentType;
};

/**
 * An item in the list of grouped items.
 */
export type Item = {
  id: string;
  title: string;
  Icon?: React.ComponentType;
  isDisabled?: boolean;
  hoverTooltip?: string;
};

export type GroupSecondaryActionConfig = {
  SecondaryActionButtonIcon: React.ComponentType<{
    group: ItemsGroup;
  }>;
  onClick(this: void, group: ItemsGroup): void;
};

export type ItemActionConfig = {
  onClick(this: void, item: Item): void;
};

export type ItemSecondaryActionConfig = {
  SecondaryActionButtonIcon: React.ComponentType<{ item: Item }>;
  onClick(this: void, item: Item): void;
};
