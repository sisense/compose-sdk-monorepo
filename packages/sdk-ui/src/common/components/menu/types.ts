import { MenuPosition, MenuItemSection } from '@/types';

export type MenuOptions = {
  position: MenuPosition;
  itemSections: MenuItemSection[];
};

export type BeforeMenuOpenHandler = (options: MenuOptions) => MenuOptions | null;
export type OpenMenuFn = (options: MenuOptions) => void;
