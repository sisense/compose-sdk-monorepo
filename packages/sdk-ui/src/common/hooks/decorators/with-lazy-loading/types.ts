import { AnyArray, AnyObject, RestApiHookState } from '@/utils/utility-types';

export type Slice = {
  count: number;
  offset: number;
};

type Sliceable = Partial<Slice>;
type Enableable = { enabled?: boolean };

export type RestApiHookParams = Sliceable & Enableable & AnyObject;

export type UnsliceableParams<Params extends RestApiHookParams> = Omit<Params, keyof Slice>;

export type LazyLoadingConfig = {
  initialCount: number;
  chunkSize: number;
  dataKey: string;
};

export type SliceableRestApiHook<
  Params extends RestApiHookParams,
  DataKey extends string,
  SuccessDataType extends AnyArray,
> = (params: Params) => RestApiHookState<DataKey, SuccessDataType>;

export type WithLoadMore<ReturnObject> = ReturnObject & {
  loadMore: () => void;
};

export type HookWithLazyLoading<
  HookParams extends RestApiHookParams,
  DataKey extends string,
  SuccessDataType extends AnyArray,
  Hook extends SliceableRestApiHook<HookParams, DataKey, SuccessDataType>,
> = (params: UnsliceableParams<HookParams>) => WithLoadMore<ReturnType<Hook>>;

export type DataChunk = {
  slice: Slice;
  data: AnyArray;
};

export type HookExecutionStatus = 'success' | 'error' | 'loading';
