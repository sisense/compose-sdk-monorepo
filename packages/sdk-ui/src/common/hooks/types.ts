export interface HookEnableParam {
  /**
   * Boolean flag to control if the hook is executed
   *
   * If not specified, the default value is `true`
   */
  enabled?: boolean;
}
