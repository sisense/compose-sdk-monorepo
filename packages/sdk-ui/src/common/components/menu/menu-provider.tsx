import { useCallback, useMemo, useState } from 'react';

import { ContextMenu } from '@/common/components/menu/context-menu';

import { MenuContext, useMenuContext } from './menu-context';
import { BeforeMenuOpenHandler, MenuOptions } from './types';

export interface MenuProviderProps {
  /** Optional handler function to process menu options before opening the menu. */
  onBeforeMenuOpen?: BeforeMenuOpenHandler;
  children?: React.ReactNode;
}

/**
 * The root-level provider component for managing and displaying the context menu.
 * This component provides the menu API for controlling a single UI menu component for the entire menu tree.
 * It also offers a middleware-like `onBeforeMenuOpen` property, which allows to intercept
 * and adjust menu options before the context menu is displayed.
 */
const MenuTreeRootProvider: React.FC<MenuProviderProps> = ({ children, onBeforeMenuOpen }) => {
  const [menuOptions, setMenuOptions] = useState<null | MenuOptions>(null);

  const openMenu = useCallback(
    (options: MenuOptions) => {
      const finalOptions = onBeforeMenuOpen ? onBeforeMenuOpen(options) : options;

      if (!finalOptions) {
        return;
      }

      setMenuOptions(finalOptions);
    },
    [onBeforeMenuOpen],
  );

  const closeMenu = useCallback(() => {
    setMenuOptions(null);
  }, []);

  const onMenuClose = useCallback(() => {
    menuOptions?.onClose?.();
    closeMenu();
  }, [closeMenu, menuOptions]);

  // Note: should have memoization to prevent redundant context update due to component state change
  const menuApi = useMemo(() => {
    return {
      openMenu,
      closeMenu,
    };
  }, [openMenu, closeMenu]);

  return (
    <MenuContext.Provider value={menuApi}>
      <ContextMenu
        position={menuOptions?.position}
        itemSections={menuOptions?.itemSections}
        closeContextMenu={onMenuClose}
        alignment={menuOptions?.alignment}
      />
      {children}
    </MenuContext.Provider>
  );
};

/**
 * The node-level provider component for managing and displaying the context menu.
 * This component provides the proxy access to the root menu API together with a possibility to set
 * a middleware-like `onBeforeMenuOpen` property which allows to intercept the nested menu sub-tree.
 */
const MenuTreeNodeProvider: React.FC<MenuProviderProps> = ({ children, onBeforeMenuOpen }) => {
  const menuSettings = useMenuContext();

  const openMenu = useCallback(
    (options: MenuOptions) => {
      const finalOptions = onBeforeMenuOpen ? onBeforeMenuOpen(options) : options;

      if (!finalOptions) {
        return;
      }

      menuSettings?.openMenu(finalOptions);
    },
    [onBeforeMenuOpen, menuSettings],
  );

  const closeMenu = useCallback(() => {
    menuSettings?.closeMenu();
  }, [menuSettings]);

  const menuApi = {
    openMenu,
    closeMenu,
  };

  return <MenuContext.Provider value={menuApi}>{children}</MenuContext.Provider>;
};

/**
 * The general menu provider component that decides whether to use the root or node-level menu provider.
 */
export const MenuProvider = ({ children, onBeforeMenuOpen }: MenuProviderProps) => {
  const menuSettings = useMenuContext();

  if (!menuSettings) {
    return (
      <MenuTreeRootProvider onBeforeMenuOpen={onBeforeMenuOpen}>{children}</MenuTreeRootProvider>
    );
  }

  if (onBeforeMenuOpen) {
    return (
      <MenuTreeNodeProvider onBeforeMenuOpen={onBeforeMenuOpen}>{children}</MenuTreeNodeProvider>
    );
  }

  return <>{children}</>;
};
