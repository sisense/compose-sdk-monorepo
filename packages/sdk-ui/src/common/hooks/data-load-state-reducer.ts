/**
 * States of a data load.
 */
export type DataState<Data> = DataLoadingState<Data> | DataErrorState | DataSuccessState<Data>;

/**
 * State of data loading.
 */
export type DataLoadingState<Data> = {
  /** Whether the data is loading */
  isLoading: true;
  /** Whether the data load has failed */
  isError: false;
  /** Whether the data load has succeeded */
  isSuccess: false;
  /** The error if any occurred */
  error: undefined;
  /** The result data if the load has succeeded */
  data: Data | undefined;
  /** The status of the data load */
  status: 'loading';
};

/**
 * State of a data load that has failed.
 */
export type DataErrorState = {
  /** Whether the data is loading */
  isLoading: false;
  /** Whether the data load has failed */
  isError: true;
  /** Whether the data load has succeeded */
  isSuccess: false;
  /** The error if any occurred */
  error: Error;
  /** The result data if the load has succeeded */
  data: undefined;
  /** The status of the data load */
  status: 'error';
};

/**
 * State of a data load that has succeeded.
 */
export type DataSuccessState<Data> = {
  /** Whether the data is loading */
  isLoading: false;
  /** Whether the data load has failed */
  isError: false;
  /** Whether the data load has succeeded */
  isSuccess: true;
  /** The error if any occurred */
  error: undefined;
  /** The result data if the load has succeeded */
  data: Data;
  /** The status of the data load */
  status: 'success';
};

export type DataLoadAction<Data> =
  | {
      type: 'loading';
    }
  | {
      type: 'success';
      data: Data;
    }
  | {
      type: 'error';
      error: Error;
    };

export function dataLoadStateReducer<Data>(
  state: DataState<Data>,
  action: DataLoadAction<Data>,
): DataState<Data> {
  switch (action.type) {
    case 'loading':
      return startLoading(state);
    case 'success':
      return applySuccessData(action.data);
    case 'error':
      return applyError(action.error);
  }
}

function startLoading<Data>(state: DataState<Data>): DataLoadingState<Data> {
  return {
    ...state,
    isLoading: true,
    isError: false,
    isSuccess: false,
    status: 'loading',
    error: undefined,
  };
}

function applySuccessData<Data>(data: Data): DataSuccessState<Data> {
  return {
    isLoading: false,
    isError: false,
    isSuccess: true,
    data,
    status: 'success',
    error: undefined,
  };
}

function applyError(error: Error): DataErrorState {
  return {
    isLoading: false,
    isError: true,
    isSuccess: false,
    data: undefined,
    status: 'error',
    error,
  };
}
