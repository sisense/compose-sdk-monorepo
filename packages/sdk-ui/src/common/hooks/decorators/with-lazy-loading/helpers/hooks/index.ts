export * from './use-data-accumulation';
export * from './use-get-load-more-function';
export * from './use-load-data-from-internal-hook';
export * from './use-manage-slice-to-load';
export * from './use-get-hook-state-result';
