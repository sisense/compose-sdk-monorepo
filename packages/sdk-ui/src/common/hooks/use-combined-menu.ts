import { useCallback, useRef } from 'react';
import { MenuOptions } from '../components/menu/types.js';
import { useMenu } from './use-menu.js';

export type CombineMenusFn = (menusOptions: MenuOptions[]) => MenuOptions;

export interface CombinedMenuParams {
  /** Function to combine the current menu options with the captured menu options */
  combineMenus: CombineMenusFn;
}

/**
 * Custom hook to manage a combined menu flow,
 * allowing to combine multiple menus options opened one after the other.
 *
 * @internal
 */
export const useCombinedMenu = ({ combineMenus }: CombinedMenuParams) => {
  const capturedMenusOptions = useRef<MenuOptions[]>([]);
  const menuApi = useMenu();

  const openCombinedMenu = useCallback(() => {
    const combinedMenusOptions =
      capturedMenusOptions.current.length === 1
        ? capturedMenusOptions.current[0]
        : combineMenus(capturedMenusOptions.current);

    menuApi.openMenu({
      ...combinedMenusOptions,
      onClose: () => {
        capturedMenusOptions.current = [];
      },
    });
  }, [menuApi, combineMenus]);

  const openMenu = useCallback(
    (options: MenuOptions) => {
      capturedMenusOptions.current.push(options);

      openCombinedMenu();
    },
    [openCombinedMenu],
  );

  return {
    openMenu,
  };
};
