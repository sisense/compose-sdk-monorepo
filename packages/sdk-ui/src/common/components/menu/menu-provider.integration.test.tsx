// integration test of MenuProvider + useCombinedMenu

import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MenuProvider } from './menu-provider';
import { useCombinedMenu } from '@/common/hooks/use-combined-menu';
import { MenuOptions } from './types';

describe('MenuProvider + useCombinedMenu', () => {
  const TestComponent = ({ menuOptions }: { menuOptions: MenuOptions }) => {
    const { openMenu } = useCombinedMenu({
      combineMenus: ([currentMenuOptions, capturedMenuOptions]) => ({
        ...currentMenuOptions,
        itemSections: currentMenuOptions.itemSections.concat(capturedMenuOptions.itemSections),
      }),
    });

    const onBeforeMenuOpen = (menuOptions: MenuOptions) => {
      openMenu(menuOptions);
      return null;
    };

    return (
      <MenuProvider onBeforeMenuOpen={onBeforeMenuOpen}>
        <div>
          <button
            data-testid="open-menu-button"
            onClick={() => {
              openMenu(menuOptions);
            }}
          />
        </div>
      </MenuProvider>
    );
  };
  it('should open the combined menu and close it when an item is clicked', async () => {
    const result = render(
      <MenuProvider>
        <TestComponent
          menuOptions={{
            position: {
              left: 42,
              top: 69,
            },
            itemSections: [{ sectionTitle: 'Some Section', items: [{ caption: 'Menu item 1' }] }],
          }}
        />
      </MenuProvider>,
    );
    const openMenuButton = result.getByTestId('open-menu-button');
    expect(openMenuButton).not.toBeNull();
    openMenuButton.click();

    await waitFor(() => {
      expect(result.getByRole('menu')).toBeInTheDocument();
      expect(result.getByText('Some Section')).toBeInTheDocument();
      expect(result.getByText('Menu item 1')).toBeInTheDocument();
    });

    //close the menu
    result.getByText('Menu item 1').click();
    await waitFor(() => {
      expect(result.queryByRole('menu')).not.toBeInTheDocument();
    });
  });

  it('should open the combined menu with the nested menu items, and close it when an item is clicked', async () => {
    const result = render(
      <MenuProvider>
        <TestComponent
          menuOptions={{
            position: {
              left: 42,
              top: 69,
            },
            itemSections: [
              {
                sectionTitle: 'Some Section',
                items: [
                  {
                    caption: 'Menu item 1',
                    subItems: [
                      {
                        sectionTitle: 'Nested Section',
                        items: [{ caption: 'Nested Menu item 1' }],
                      },
                    ],
                  },
                ],
              },
            ],
          }}
        />
      </MenuProvider>,
    );
    const openMenuButton = result.getByTestId('open-menu-button');
    expect(openMenuButton).not.toBeNull();
    openMenuButton.click();

    await waitFor(() => {
      expect(result.getByRole('menu')).toBeInTheDocument();
      expect(result.getByText('Some Section')).toBeInTheDocument();
      expect(result.getByText('Menu item 1')).toBeInTheDocument();
    });

    // hover on 'Menu item 1' to open the nested menu
    await userEvent.hover(result.getByText('Menu item 1'));

    await waitFor(() => {
      // nested menu should be open together with the parent menu
      expect(result.getByRole('menu')).toBeInTheDocument();
      expect(result.getByText('Some Section')).toBeInTheDocument();
      expect(result.getByText('Menu item 1')).toBeInTheDocument();
      expect(result.getByText('Nested Section')).toBeInTheDocument();
      expect(result.getByText('Nested Menu item 1')).toBeInTheDocument();
    });

    // re-hover on 'Nested Menu item 1'
    await userEvent.hover(result.getByText('Nested Menu item 1'));
    await waitFor(() => {
      // nested menu should still be open
      expect(result.getByText('Nested Section')).toBeInTheDocument();
      expect(result.getByText('Nested Menu item 1')).toBeInTheDocument();
    });

    // unhover on 'Menu item 1' to hide the nested menu
    await userEvent.unhover(result.getByText('Nested Menu item 1'));
    await waitFor(() => {
      // nested menu should be closed, original menu should still be open
      expect(result.queryByText('Nested Section')).not.toBeInTheDocument();
      expect(result.queryByText('Nested Menu item 1')).not.toBeInTheDocument();
    });

    // hover again on 'Menu item 1' to open the nested menu
    await userEvent.hover(result.getByText('Menu item 1'));
    await waitFor(() => {
      // nested menu should be open again
      expect(result.getByText('Nested Section')).toBeInTheDocument();
      expect(result.getByText('Nested Menu item 1')).toBeInTheDocument();
    });
    // unhover on 'Menu item 1' to hide the nested menu
    await userEvent.unhover(result.getByText('Menu item 1'));
    await waitFor(() => {
      // nested menu should be closed, original menu should still be open
      expect(result.getByRole('menu')).toBeInTheDocument();
      expect(result.getByText('Some Section')).toBeInTheDocument();
      expect(result.getByText('Menu item 1')).toBeInTheDocument();
      expect(result.queryByText('Nested Section')).not.toBeInTheDocument();
      expect(result.queryByText('Nested Menu item 1')).not.toBeInTheDocument();
    });

    // hover again on 'Menu item 1' to open the nested menu
    // click on 'Nested Menu item 1' to close the menu
    await userEvent.hover(result.getByText('Menu item 1'));
    result.getByText('Nested Menu item 1').click();
    await waitFor(() => {
      expect(result.queryByRole('menu')).not.toBeInTheDocument();
    });
  });
});
