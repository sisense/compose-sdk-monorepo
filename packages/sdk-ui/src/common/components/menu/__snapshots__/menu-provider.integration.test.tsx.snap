// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MenuProvider + useCombinedMenu > should open the combined menu and close it when an item is clicked 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div>
      <button
        data-testid="open-menu-button"
      />
    </div>
  </div>
  <div
    class="MuiPopover-root MuiMenu-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper MuiMenu-paper MuiMenu-paper css-2ie444-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
      style="opacity: 1; transform: none; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 69px; left: 42px; transform-origin: 0px 0px;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiList-padding MuiList-dense MuiMenu-list css-1sf1hir-MuiList-root-MuiMenu-list"
        role="menu"
        tabindex="-1"
      >
        <div
          class="csdk-menu-section"
          tabindex="0"
        >
          <li
            class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-h2pzkj-MuiListSubheader-root"
          >
            Some Section
          </li>
          <div
            class="csdk-menu-item "
          >
            <li
              class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-11utrpy-MuiButtonBase-root-MuiMenuItem-root"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="csdk-menu-item-content"
              >
                Menu item 1
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </li>
          </div>
        </div>
      </ul>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`MenuProvider + useCombinedMenu > should open the combined menu with the nested menu items, and close it when an item is clicked 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div>
      <button
        data-testid="open-menu-button"
      />
    </div>
  </div>
  <div
    class="MuiPopover-root MuiMenu-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper MuiMenu-paper MuiMenu-paper css-2ie444-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
      style="opacity: 1; transform: none; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 69px; left: 42px; transform-origin: 0px 0px;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiList-padding MuiList-dense MuiMenu-list css-1sf1hir-MuiList-root-MuiMenu-list"
        role="menu"
        tabindex="-1"
      >
        <div
          class="csdk-menu-section"
          tabindex="0"
        >
          <li
            class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-h2pzkj-MuiListSubheader-root"
          >
            Some Section
          </li>
          <div
            class="csdk-menu-item "
          >
            <li
              class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-upema-MuiButtonBase-root-MuiMenuItem-root"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="csdk-menu-item-content"
                style="width: 100%; display: flex; align-items: center; justify-content: space-between;"
              >
                Menu item 1
                <svg
                  height="24px"
                  viewBox="0 0 24 24"
                  width="24px"
                >
                  <path
                    d="M13.34 12.5l-2.72-3.175a.5.5 0 1 1 .76-.65l2.998 3.5a.5.5 0 0 1 0 .65l-2.998 3.5a.5.5 0 1 1-.76-.65l2.72-3.175z"
                    fill="#5B6372"
                  />
                </svg>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </li>
          </div>
        </div>
      </ul>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`MenuProvider + useCombinedMenu > should open the combined menu with the nested menu items, and close it when an item is clicked 2`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div>
      <button
        data-testid="open-menu-button"
      />
    </div>
  </div>
  <div
    aria-hidden="true"
    class="MuiPopover-root MuiMenu-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper MuiMenu-paper MuiMenu-paper css-2ie444-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
      style="opacity: 1; transform: none; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 69px; left: 42px; transform-origin: 0px 0px;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiList-padding MuiList-dense MuiMenu-list css-1sf1hir-MuiList-root-MuiMenu-list"
        role="menu"
        tabindex="-1"
      >
        <div
          class="csdk-menu-section"
          tabindex="0"
        >
          <li
            class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-h2pzkj-MuiListSubheader-root"
          >
            Some Section
          </li>
          <div
            class="csdk-menu-item "
          >
            <li
              class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-upema-MuiButtonBase-root-MuiMenuItem-root"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="csdk-menu-item-content"
                style="width: 100%; display: flex; align-items: center; justify-content: space-between;"
              >
                Menu item 1
                <svg
                  height="24px"
                  viewBox="0 0 24 24"
                  width="24px"
                >
                  <path
                    d="M13.34 12.5l-2.72-3.175a.5.5 0 1 1 .76-.65l2.998 3.5a.5.5 0 0 1 0 .65l-2.998 3.5a.5.5 0 1 1-.76-.65l2.72-3.175z"
                    fill="#5B6372"
                  />
                </svg>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </li>
          </div>
        </div>
      </ul>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
  <div
    class="MuiPopover-root MuiMenu-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
    role="presentation"
    style="pointer-events: none;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper MuiMenu-paper MuiMenu-paper css-2ie444-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
      style="opacity: 1; transform: none; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiList-padding MuiMenu-list css-1exmg4n-MuiList-root-MuiMenu-list"
        role="menu"
        tabindex="-1"
      >
        <div
          style="pointer-events: auto;"
          tabindex="0"
        >
          <div
            class="csdk-menu-section"
          >
            <li
              class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-h2pzkj-MuiListSubheader-root"
            >
              Nested Section
            </li>
            <div
              class="csdk-menu-item "
            >
              <li
                class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-vj38tt-MuiButtonBase-root-MuiMenuItem-root"
                role="menuitem"
                tabindex="-1"
              >
                <div
                  class="csdk-menu-item-content"
                >
                  Nested Menu item 1
                </div>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </li>
            </div>
          </div>
        </div>
      </ul>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
